### ISSUE TITLE
What is the expected gpu memory performance drop wrt flash attention with block masks?

### ISSUE THREAD

**Issue by arilato** [pytorch-labs/attention-gym#54]
I'm testing out using flex attention to utilize some custom attention masks. The attention masks I'm working with are causal, except there is usually a relatively small single rectangular area of 0s in the mask, which I added support for with a block mask.

Previously, using xformers which maps to flash attention 2, I managed to train mistral-large at 28k sequence length on 8xh100. However, with flex attention, even 24k sequence length is running OOM. Is thie expected?

I am compiling `flex_attention` after importing it, and also compiling the block mask when instantiating it.

**Comment by Chillee** [comment]
@arilato I would not expect that. The extra memory overhead from FlexAttention should be S^2/(BLOCK_SIZE^2). With the default BLOCK_SIZE of 128, at 28k, the extra memory overhead should be around 100kb or so.

**Comment by drisspg** [comment]
@arilato you need to also ensure that you are compiling the `create_block_mask` function since if this is called without compile it will realized the full 28k x 28k sequence tensor

**Comment by ViktorooReps** [comment]
For some reason I see more memory consumption compared to SDPA even with compilation:

![output_mem](https://github.com/user-attachments/assets/d102e83b-8cc8-45da-8053-e8d3879b8322)

But I have other issues: SDPA breaks at 60k context length (see [#pytorch](https://github.com/pytorch/pytorch/issues/133976#issue-2474822007)) and no performance improvements when causal mask is changed to sliding window: #83
